core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -0,0 +1,783 @@
+#define QM_SQC_VFT_SQ_NUM_SHIFT 45
+#define QM_CQC_VFT_VALID BIT(28)
+#define QM_VFT_WRITE 0
+#define QM_VFT_READ 1
+#define QM_SQC_VFT_BASE_MASK 0x3ff
+#define QM_SQC_VFT_NUM_MASK 0x3ff
+/* QM INIT */
+#define QM_MEM_START_INIT 0x100040
+#define QM_MEM_INIT_DONE 0x100044
+#define QM_VF_AEQ_INT_MASK 0x4
+#define QM_VF_EQ_INT_MASK 0xc
+#define QM_ARUSER_M_CFG_1 0x100088
+#define QM_ARUSER_M_CFG_ENABLE 0x100090
+#define QM_AWUSER_M_CFG_1 0x100098
+#define QM_AWUSER_M_CFG_ENABLE 0x1000a0
+#define QM_AXUSER_CFG 0x40001070
+#define AXUSER_M_CFG_ENABLE 0x7ffffc
+#define QM_AXI_M_CFG 0x1000ac
+#define AXI_M_CFG 0xffff
+#define QM_PEH_AXUSER_CFG 0x1000cc
+#define PEH_AXUSER_CFG 0x400801
+#define QM_CACHE_CTL 0x100050
+#define QM_CACHE_CFG 0x4893
+#define QM_CACHE_WB_START 0x100204
+#define QM_CACHE_WB_DONE 0x100208
+/* XQC shift */
+#define QM_SQ_SQE_SIZE_SHIFT 12
+#define QM_SQ_ORDER_SHIFT 4
+#define QM_SQ_TYPE_SHIFT 8
+#define QM_CQE_SIZE 4
+#define QM_CQ_CQE_SIZE_SHIFT 12
+/* CQE */
+#define QM_CQE_PHASE(cqe) (((cqe)->w7) & 0x1)
+
+enum qm_mailbox_common_cmd {
+	QM_MB_CMD_SQC = 0x0,
+	QM_MB_CMD_CQC,
+	QM_MB_CMD_EQC,
+	QM_MB_CMD_AEQC,
+	QM_MB_CMD_SQC_BT,
+	QM_MB_CMD_CQC_BT,
+	QM_MB_CMD_SQC_VFT,
+};
+
+enum qm_mailbox_cmd_v3 {
+	QM_MB_CM_CLOSE_QM = 0x7,
+	QM_MB_CMD_CLOSE_QP,
+	QM_MB_CMD_FLUSH_QM,
+	QM_MB_CMD_FLUSH_QP,
+	QM_MB_CMD_SRC = 0xc,
+	QM_MB_CMD_DST,
+	QM_MB_CMD_STOP_QM,
+};
+
+struct qm_mailbox {
+	uint16_t w0;
+	uint16_t queue;
+	uint32_t base_l;
+	uint32_t base_h;
+	uint32_t token;
+};
+
+struct qm_dfx_registers {
+	const char *reg_name;
+	uint32_t reg_offset;
+};
+
+static struct qm_dfx_registers qm_dfx_regs[] = {
+	{"QM_ECC_1BIT_CNT           ",  0x104000},
+	{"QM_ECC_MBIT_CNT           ",  0x104008},
+	{"QM_DFX_MB_CNT             ",  0x104018},
+	{"QM_DFX_DB_CNT             ",  0x104028},
+	{"QM_DFX_SQE_CNT            ",  0x104038},
+	{"QM_DFX_CQE_CNT            ",  0x104048},
+	{"QM_DFX_SEND_SQE_TO_ACC_CNT",  0x104050},
+	{"QM_DFX_WB_SQE_FROM_ACC_CNT",  0x104058},
+	{"QM_DFX_ACC_FINISH_CNT     ",  0x104060},
+	{"QM_DFX_CQE_ERR_CNT        ",  0x1040b4},
+	{ NULL, 0}
+};
+
+void hisi_qm_get_version(struct hisi_qm *qm)
+{
+	uint32_t val = 0;
+
+	val = io_read32(qm->io_base + QM_REVISON_ID_BASE);
+	qm->version = val & QM_REVISON_ID_MASK;
+}
+
+static void qm_db(struct hisi_qm *qm, uint16_t qn, uint8_t cmd, uint16_t index,
+		  uint8_t priority)
+{
+	uint16_t rand_data = QM_DB_RAND_DATA;
+	uint64_t doorbell = 0;
+
+	doorbell = (qn | ((uint64_t)cmd << QM_DB_CMD_SHIFT) |
+		   ((uint64_t)rand_data << QM_DB_RAND_DATA_SHIFT) |
+		   ((uint64_t)index << QM_DB_INDEX_SHIFT) |
+		   ((uint64_t)priority << QM_DB_PRIORITY_SHIFT));
+
+	io_write64(qm->io_base + QM_DOORBELL_SQ_CQ_BASE, doorbell);
+}
+
+static int32_t qm_wait_mb_ready(struct hisi_qm *qm)
+{
+	uint32_t val = 0;
+
+	/* return 0 mailbox ready, -ETIMEDOUT hardware timeout */
+	return readl_relaxed_poll_timeout(qm->io_base + QM_MAILBOX_BASE,
+					  val, !((val >> QM_MB_BUSY_SHIFT) &
```

Comment:
> I'd rather define `QM_MB_BUSY_BIT` as `BIT32(QM_MB_BUSY_SHIFT)` and then write it as: `!(val & QM_MB_BUSY_BIT)`. It looks like a few other places could benefit from `QM_MB_BUSY_BIT` instead of open-coding the shifting.

Good point.